### ISSUE TITLE
App Actions (Quick Actions/App Shortcuts)

### ISSUE THREAD

**Issue by jamesmontemagno** [xamarin/Essentials#148]
### Description

API for iOS's [Quick Actions](https://developer.apple.com/ios/human-interface-guidelines/extensions/home-screen-actions) and Android's [App Shortcuts](https://developer.android.com/guide/topics/ui/shortcuts.html)

API:
```
event OnAppAction(string actionType);
IEnumerable<AppActionItem> ActionItems {get;set;}
```

AppActionItem
```
string ActionType {get; set;}
string LocalizedTitle {get;set;}
string Icon {get;set;} 
```

On iOS we should pull from xcassets or resources and drawables on Android.

Event must be registered early enough is application lifecycle to handle events. (Application or AppDelegate).

**Comment by JakeStaTeresa** [comment]
@jamesmontemagno App Shortcuts link points to the same link as Quick Actions. It should point to https://developer.android.com/guide/topics/ui/shortcuts.html

**Comment by Redth** [comment]
Community contributions are welcome on this one!

**Comment by edsnider** [comment]
I have the iOS and Android code for this and can contribute... just need to think through the `OnAppAction` event. The way the "actions" come into the application would require the user to wire up something that could then invoke the `OnAppAction`. Doesn't feel very clean though.  Looks like @Redth is also dealing with this in the Local Notifications. Thoughts?

**Comment by newky2k** [comment]
@edsnider  did you get anywhere with this?

**Comment by edsnider** [comment]
@newky2k I was waiting for it to come back out of “Needs triage” before I pushed anything.

**Comment by heytherewill** [comment]
A think a good approach for this API is to have only two methods:

`Task SetShortcuts(Shortcut[] shortcuts)`
This will replace all existing shortcuts with the ones provided in the array (or whichever collection type we settle to use). 

`Task ClearShortcuts()`
Shorthand for `SetShortcuts(new Shortcut[0])`

This much API surface should cover most/all needs of developers and is really straight forward to both implement and maintain. I recommend using `Task` over void because iOS requires Quick Actions to be set on the main thread.

Once we settle on the specs, if @edsnider is for whatever reason unable to work on this, I'd love to contribute to this API. In either case, we have implemented a similar API for the Toggl mobile apps, so maybe the code can be useful to whoever implements this. The code is available for [Android](https://github.com/toggl/mobileapp/blob/develop/Toggl.Droid/ApplicationShortcutCreator.cs) and [iOS](https://github.com/toggl/mobileapp/blob/develop/Toggl.iOS/ApplicationShortcutCreator.cs)

**Comment by manishkungwani** [comment]
@newky2k I'll try to implement this and come back with details.

**Comment by mattleibow** [comment]
Merged in  #1190
